### PR TITLE
Implement Element.insertAdjacentHTML()

### PR DESCRIPTION
**Pull request by kasperisager**
Ref. #1219. Grabbed and slightly tweaked the implementation from https://github.com/assaf/zombie/blob/005b4726c9c8bd02a234cb26cb756d90358bf0a4/src/dom/jsdom_patches.js#L60-L90

A test case has been added, but both this as well as the implementation might need some more work. It's a start though!

### CODE REVIEW

**Comment by kasperisager** [comment]
Also, the WHATWG spec doesn't yet contain a definition of insertAdjacentHTML but I've added a reference to https://dom.spec.whatwg.org/#dom-element-insertadjacenthtml as I wasn't sure what else to put there (https://w3c.github.io/DOM-Parsing/#widl-Element-insertAdjacentHTML-void-DOMString-position-DOMString-text maybe?).

**Comment by domenic** [comment]
Ah excellent! Could you enable https://github.com/w3c/web-platform-tests/blob/master/domparsing/insert-adjacent.html as well (and make sure it passes)? And probably move your test into to-upstream/domparsing.

**Comment by kasperisager** [comment]
Ah, I totally missed that one! Should make my test case obsolete then.

File: lib/jsdom/living/nodes/Element-impl.js [hunk 1]
```diff
@@ -430,6 +430,43 @@ class ElementImpl extends NodeImpl {
 
     return attr;
   }
+
+  insertAdjacentHTML(position, text) {
+    const parentNode = this.parentNode;
+    const container = this.ownerDocument.createElementNS("http://www.w3.org/1999/xhtml", "_");
```
Comment (domenic) [inline lib/jsdom/living/nodes/Element-impl.js]: You should use `this.ownerDocument.createElement("template")` as the container. Also you should add some tests showing how using `"_"` fails for elements like `thead`, `tr`, `td`.

**Comment by domenic** [comment]
The spec is actually at https://w3c.github.io/DOM-Parsing/#widl-Element-insertAdjacentHTML-void-DOMString-position-DOMString-text

File: lib/jsdom/living/nodes/Element-impl.js [hunk 2]
```diff
@@ -430,6 +430,43 @@ class ElementImpl extends NodeImpl {
 
     return attr;
   }
+
+  insertAdjacentHTML(position, text) {
```
Comment (domenic) [inline lib/jsdom/living/nodes/Element-impl.js]: Looks like "If context is null or a document, throw a DOMException with name "NoModificationAllowedError"." would be good to add a test for this too.

File: lib/jsdom/living/nodes/Element-impl.js (hunk 1 above)
Comment (domenic) [inline lib/jsdom/living/nodes/Element-impl.js]: Hmm reading the spec it seems it does something a bit more complicated, e.g. defaulting to body in certain cases. I think it's OK to just use template as it will be most correct, but be careful and check browsers. If there are some edge cases that neither "template" nor "body" cover, then it's OK, we can just leave a TODO.

**Comment by domenic** [comment]
FYI the existing web platform test is failing; not sure exactly why...

**Comment by kasperisager** [comment]
Read through the spec more thoroughly and re-did what I had copied from zombie. The tests pass on my end now :+1:

**Comment by kasperisager** [comment]
I should note that not all conditions of step 2 are checked.

File: lib/jsdom/living/nodes/Element-impl.js [hunk 3]
```diff
@@ -430,6 +430,58 @@ class ElementImpl extends NodeImpl {
...
+    position = position.toLowerCase();
+
+    let context;
```
Comment (domenic) [inline lib/jsdom/living/nodes/Element-impl.js]: So as-is context is unused. I think the spec is not going to be very useful for a guidance on this since it includes "invoking the fragment parsing algorithm with text as markup, and context as the context element" which I don't think we have the ability to do (yet). I made a note of that in https://github.com/tmpvar/jsdom/pull/1316#issuecomment-172030345

Instead I think you should remove context, and just check this.parentNode in line 442. Lines 542 and 460-462 can go away.

File: lib/jsdom/living/nodes/Element-impl.js [hunk 4]
```diff
@@ -430,6 +430,58 @@ class ElementImpl extends NodeImpl {
...
+
+    switch (position) {
+      case "beforebegin":
+      case "afterend":
+        context = this.parentNode;
+
+        if (context === null || context.nodeType === NODE_TYPE.DOCUMENT_NODE) {
+          throw new DOMException(DOMException.NO_MODIFICATION_ALLOWED_ERR, "Modifications are not allowed for this" +
```
Comment (domenic) [inline lib/jsdom/living/nodes/Element-impl.js]: I don't believe the existing test suite covers these exceptions, so adding to-upstream tests for them would be lovely.

**Comment by domenic** [comment]
Thank you very much; lovely work! Merged as 33b93417f93c41e67ba01e95faedfb1d542fdff7.

**Comment by kasperisager** [comment]
Cheers!

**Comment by RReverser** [comment]
Looks like this should be trivial to extend to support `insertAdjacentText`?

**Comment by Zirro** [comment]
@RReverser We would certainly appreciate a pull request doing so, assuming it passes the relevant tests.
